### PR TITLE
Improve access speed of sensor values by caching attribute files

### LINKED ISSUE #93: [question] Go seven times slower in retrieving sensor values than Python?
I recently tried Go on the EV3 because I hoped it would have better performance compared to Python.

In a while loop, the robot repeatedly retrieves a sensor value (in this case, from mindsensors.com Line Leader) and thus waits until it changes to a specific value.
I converted the corresponding Python code to Go, but I noticed Go is actually seven times slower in retrieving those sensor values. Below are the two test programs I wrote to compare the performance.

```python
ll = Sensor("in2")
ll.mode = "PID-ALL"
debug_print("Starting")
start = time.perf_counter()
for _ in range(1000):
    value = ll.value(2)
    if value == -1:
        # is on black line; the program would normally end the loop here
        pass
end = time.perf_counter()-start
debug_print(end)
```
Output (example): `0.9874834319999763`
```go
func main() {
    ll, err := ev3dev.SensorFor("ev3-ports:in2:i2c1", "ms-line-leader")
    if err != nil {
        log.Fatal(err)
    }
    ll.SetMode("PID-ALL")

    log.Println("Starting")
    start := time.Now()
    for i := 0; i < 1000; i++ {
        value, err := ll.Value(2)
        if err != nil {
            log.Fatal(err)
        }
        if value == "-1" {
            // is on black line
        }
    }
    end := time.Since(start)
    log.Println(end)
}
```
Output (example): `6.850100872s`

While Python takes about 1 second on my EV3 for the loop, Go takes about 7 seconds.

The question I'm asking myself is: Am I doing something wrong here? Because I usually would expect Go to be faster, and indeed the startup time is much lower.
If not: Why is Go seven times slower? I couldn't find any major difference between how the python `ev3dev2` module and the Go `github.com/ev3go/ev3dev` package retrieve sensor values.

PS: I think ~143 checks per second with Go are more than enough, but I'm opening this issue because I'm curious to find out what causes  this. As you might have guessed, I've got little to no experience with Go.

### PR DESCRIPTION
**Pull request by FlorianRaediker**
Fixes #93.

I also replaced Sensor.BinData and attributeOf with the new function binAttributeOf so that BinData() also uses the cached files.

### CODE REVIEW

File: ev3dev.go [hunk 1]
```diff
@@ -522,17 +522,35 @@ func (d byID) Len() int           { return len(d) }
 func (d byID) Less(i, j int) bool { return d[i].id < d[j].id }
 func (d byID) Swap(i, j int)      { d[i], d[j] = d[j], d[i] }
 
-func attributeOf(d Device, attr string) (dev Device, data string, _attr string, err error) {
+var openedFiles = map[string]*os.File{}
```
Comment (kortschak) [inline ev3dev.go:525]: This isn't concurrency safe.

File: ev3dev.go [hunk 2]
```diff
@@ -522,17 +522,35 @@ func (d byID) Len() int           { return len(d) }
...
+		file, err = os.Open(path)
+		if err != nil {
+			return d, nil, "", newAttrOpError(d, attr, "", "read", err)
+		}
+		openedFiles[path] = file
+	} else {
+		file = openedFiles[path]
+		file.Seek(0, 0)
+	}
+	b, err := ioutil.ReadAll(file)
```
Comment (kortschak) [inline ev3dev.go:544]: Same here.

**Review by kortschak — COMMENTED** [review]
There are some subtleties that need to be considered wrt concurrency. Also, I would have really just replaced `ioutil.ReadFile` with the registry-aware equivalent rather than putting all the logic into the file reading.

**Comment by kortschak** [comment]
Please try something like this:
```diff
diff --git a/ev3dev.go b/ev3dev.go
index 31da9dd..8186270 100644
--- a/ev3dev.go
+++ b/ev3dev.go
@@ -528,7 +528,7 @@ func attributeOf(d Device, attr string) (dev Device, data string, _attr string,
                return d, "", "", err
        }
        path := filepath.Join(d.Path(), d.String(), attr)
-       b, err := ioutil.ReadFile(path)
+       b, err := readFile(path)
        if err != nil {
                return d, "", "", newAttrOpError(d, attr, string(b), "read", err)
        }
@@ -633,3 +633,36 @@ func setAttributeOf(d Device, attr, data string) error {
        }
        return nil
 }
+
+var (
+       fileRegLock sync.Mutex
+       files       = make(map[string]*os.File)
+)
+
+func readFile(path string) ([]byte, error) {
+       f, err := fileFor(path)
+       if err != nil {
+               return nil, err
+       }
+       var buf [4096]byte // TODO(kortschak): Investigate reducing the size of the buffer.
+       n, err := f.ReadAt(buf[:], 0)
+       if err == io.EOF {
+               err = nil
+       }
+       return buf[:n], err
+}
+
+func fileFor(path string) (*os.File, error) {
+       defer fileRegLock.Unlock()
+       fileRegLock.Lock()
+       f, ok := files[path]
+       if ok {
+               return f, nil
+       }
+       f, err := os.Open(path)
+       if err != nil {
+               return nil, err
+       }
+       files[path] = f
+       return f, nil
+}
diff --git a/sensor.go b/sensor.go
index 45a8e6b..4738048 100644
--- a/sensor.go
+++ b/sensor.go
@@ -5,7 +5,6 @@
 package ev3dev
 
 import (
-       "io/ioutil"
        "os"
        "path/filepath"
        "strconv"
@@ -126,7 +125,7 @@ func (s *Sensor) BinData() ([]byte, error) {
                return nil, err
        }
        path := filepath.Join(s.Path(), s.String(), binData)
-       b, err := ioutil.ReadFile(path)
+       b, err := readFile(path)
        if err != nil {
                return nil, newAttrOpError(s, binData, string(b), "read", err)
        }
```

This fails local tests (as the PR here does on travis) because the current testing has an assumption of statelessness in the ev3go/ev3dev code, with state being held only in the sysfs. If this works on the device, I can put some work into making the test suite work.

**Comment by FlorianRaediker** [comment]
It works on my EV3. Using a buffer size of 128 is twice as fast. I measured the following durations for 10,000 LineLeader value checks:

- Python with ev3dev2: ~16s
- Python with [ev3fast](https://github.com/QuirkyCort/ev3dev-lang-python-fast): ~10s
- Go (no changes): ~40s
- Go (my code):  ~10s
- Go (your code, 4096 buffer size): ~10s
- Go (your code, 128 buffer size): ~6s

**Comment by kortschak** [comment]
OK. So then I need to figure out how to get this to be acceptable for the test suite then.

**Comment by FlorianRaediker** [comment]
> OK. So then I need to figure out how to get this to be acceptable for the test suite then.

Thanks.

**Pull request closed without merging**